// exmes/build/src/inputStream.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const pull = require('pull-stream');
const split = require('pull-split');
const file = require('pull-file');
class FileInputStream {
    constructor(path, separator, bufferSize = 100) {
        this.path = path;
        this.separator = separator;
        this.bufferSize = bufferSize;
        this.get = pull(file(path, { bufferSize }), split(separator));
    }
    fulfillWithResult(endResult, resolve, reject) {
        if (endResult === true) {
            return resolve(null);
        }
        return reject(endResult);
    }
    read() {
        return new Promise((resolve, reject) => {
            if (this.cachedEndResult) {
                return this.fulfillWithResult(this.cachedEndResult, resolve, reject);
            }
            this.get(null, (endResult, data) => {
                if (endResult) {
                    this.cachedEndResult = endResult;
                    return this.fulfillWithResult(endResult, resolve, reject);
                }
                return resolve(data);
            });
        });
    }
}
exports.FileInputStream = FileInputStream;
