// exmes/build/src/outputStream.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = require("fs");
const inputStream_1 = require("./inputStream");
class FileOutputStream {
    constructor(path, separator) {
        this.path = path;
        this.separator = separator;
        this.firstWrite = true;
        this.writable = fs_1.createWriteStream(path);
    }
    write(value) {
        return new Promise((resolve) => {
            const prefix = this.firstWrite ? '' : this.separator;
            this.writable.write(prefix + value, resolve);
            this.firstWrite = false;
        });
    }
    batchWrite(values) {
        return new Promise((resolve) => {
            const data = values.join(this.separator);
            this.writable.write(data, resolve);
            this.firstWrite = false;
        });
    }
    end() {
        this.writable.end();
    }
    getInputStream() {
        return new inputStream_1.FileInputStream(this.path, this.separator);
    }
}
exports.FileOutputStream = FileOutputStream;
